CRI: Add init containers

feiskyer opened the pull request:
This PR adds init containers support in CRI.

CC @yujuhong @Random-Liu @yifan-gu 

Also CC @kubernetes/sig-node @kubernetes/sig-rktnetes

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33363)

<!-- Reviewable:end -->

feiskyer committed:
CRI: Add init containers

yujuhong commented:
@feiskyer have you run the init container tests (using node e2e)?

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-:-KScJaJ1wEnBI64fGwAb:b-40fgbi)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 450 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KScIWbFLT6dtYtsrUDl:-KScIWbFLT6dtYtsrUDm:bmq1a52) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L450)):*
> ```Go
> 				message := fmt.Sprintf("Failed to initialize pod. %q will be restarted.", container.Name)
> 				glog.V(1).Info(message)
> 				changes.ContainersToStart[index] = message
> ```

I know you copied this from `docker_manager.go`, and there was probably a good reason for having this logic, I just couldn't remember why there would be containers running if initialization failed. 

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 478 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KScFN4JvRvmDum_VatR:-KScFN4JvRvmDum_VatS:b-a4vpq5) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L478)):*
> ```Go
> 
> 	// Don't keep init containers if they are the only containers to keep.
> 	if !sandboxChanged && len(changes.ContainersToStart) == 0 && len(changes.ContainersToKeep) == 0 {
> ```

Why would this happen?

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 482 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KScG983O2uRWHIErbCw:-KScG983O2uRWHIErbCx:bplreeo) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L482)):*
> ```Go
> 	}
> 
> 	// compute containers that to be killed
> ```

remove "that"

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363)*
<!-- Sent from Reviewable.io -->

feiskyer commented:
> @feiskyer have you run the init container tests (using node e2e)?

Yes, all init container tests are passed.

euank commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -489,6 +490,14 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *api.Pod, r
 					}
 				}
+				if containerSpec == nil {
+					for i, c := range pod.Spec.InitContainers {
```
Since this for loop is duplicated verbatim, maybe it's worth having a `func (containerlist) getContainerByName(name) *container` helper. Fine either way, just saves some indentation.

feiskyer replied:
sure.

euank commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -507,6 +516,83 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *api.Pod, r
+		initContainerNames.Insert(container.Name)
+	}
+	for name := range initContainerNames {
+		count := 0
```
nit, `bool isFirst` might be clearer here. I don't think that logging the count below (only place it's used where a bool would change things) actually matters, cc @yujuhong

feiskyer replied:
hmm, the count is just for clear. I'm ok with this. @yujuhong what do you think?

euank commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -507,6 +516,83 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *api.Pod, r
+
+	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
+		container := &pod.Spec.InitContainers[i]
+		status := podStatus.FindContainerStatusByName(container.Name)
```
I'd rather have the `if status == nil continue` logic here, not as part of the switch.

That also lets the switch become `switch status.State`

feiskyer replied:
sure.

euank commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -507,6 +516,83 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *api.Pod, r
+			case status.ExitCode != 0:
+				return &pod.Spec.InitContainers[i], status, false
+			// all init containers successful
+			case i == (len(pod.Spec.InitContainers) - 1):
```
By the comment, I'd expect `i==0` here.

Let's say you have a pod with two init containers that look like:

`[{name: foo, exitcode: 1, status: exited}, {name: bar, exitcode 0, status: exited}]`

the loop starts at `i = 1`, which leads to it checking bar (exited, exit success) and then this check succeeding, and thus it returning `done = true` even though `foo` failed so it should retry.

cc @smarterclayton am I missing something? I double checked and this is copied verbatim from dockertools, so I suspect it might be an existing bug, but haven't tried to experimentally reproduce it.

feiskyer replied:
> [{name: foo, exitcode: 1, status: exited}, {name: bar, exitcode 0, status: exited}]

That won't happen. Next init container only runs after the previous one is success.

smarterclayton replied:
There should be an earlier loop that if any init container exists as failed flags the whole init container / pod for restart / failure (while processing the pod in syncPod and determining which containers to keep, the first failed pod is sufficient to prevent all subsequent containers from being started).

If that guard is not there, not effective, or was never there, your scenario could happen, but I think we would want the guard and the rely on the correct logic here.

euank replied:
I guess that works then, but it still feels a bit unclear. At the very least, a comment to that effect would be nice.

yifan-gu replied:
I agree this part of logic is not easy to parse. I am thinking about ways to improve it.

feiskyer replied:
@euank @smarterclayton Made a little refactor, PTAL.

euank commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -380,14 +387,40 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *api.Pod, pod
+	Containers:
+		// keep all successfully completed containers up to and including the first failing container
+		for i, container := range pod.Spec.InitContainers {
+			containerStatus := podStatus.FindContainerStatusByName(container.Name)
```
nilcheck and continue here, again lets the switch be specifically on containerStatus.State

feiskyer replied:
ack

euank commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -380,14 +387,40 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *api.Pod, pod
+				changes.InitContainersToKeep[containerStatus.ID] = i
+			case containerStatus.State == kubecontainer.ContainerStateExited:
+				changes.InitContainersToKeep[containerStatus.ID] = i
+				if containerStatus.ExitCode != 0 {
```
I think this little chunk of code might be easier to read as a "containerStatusFailed" helper it it doesn't already exist.

The logic just boils down to: `if containerStatusFailed(container) { initFailed = true; break } else { containersToKeep = i }` if we have the helper I think, and it also saves us from needing a break label.

feiskyer replied:
well, container should be kept event failed, so the logic will be same of changing it from `switch` to `if..else`

euank commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -585,7 +640,61 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *api.Pod, _ api.PodStatus, podSt
-	// Step 5: start containers in podContainerChanges.ContainersToStart.
+	// Step 5: start init containers.
+	next, status, done := findActiveInitContainer(pod, podStatus)
+	if status != nil {
```
nit: `status != nil && status.ExitCode != 0` saves an indentation and I don't think it's less readable

feiskyer replied:
ack.

feiskyer commented:
>   if !sandboxChanged && len(changes.ContainersToStart) == 0 && len(changes.ContainersToKeep) == 0 {

@yujuhong I think it doesn't happen normally, but be afraid of missing some scenarios.

yujuhong commented:
*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 530 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSLPVZOlpG1huhvy-bF-r1-530:-KShC-qfONfDmPw6Kwt8:b-hiax4i) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_container.go#L530)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> hmm, the count is just for clear. I'm ok with this. @yujuhong what do you think?
</details>
Given that we print `count` in the log message below, let's just keep it as it is. @euank are you ok with that? 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
Review status: 0 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 572 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSLPVZOlpG1huhvy-bF-r1-572:-KShD2te2SP7Mvc9QLwH:b-pbdqz3) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_container.go#L572)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> sure.
</details>
Agreed.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363)*
<!-- Sent from Reviewable.io -->

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -560,6 +557,84 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *api.Pod, r
+	for name := range initContainerNames {
+		count := 0
+		for _, status := range podStatus.ContainerStatuses {
+			if status.Name != name || !initContainerNames.Has(status.Name) || status.State != kubecontainer.ContainerStateExited {
```
Why `!initContainerNames.Has(status.Name)` is needed?

yifan-gu replied:
Maybe we should make `kubecontainer.Status` to have something like `InitContainerStatus` to make our life easier ? @yujuhong @feiskyer

euank replied:
I'm not sure I understand what you mean by that. How would it make things easier?

feiskyer replied:
Not sure whether that could make things easier, same logic is required while populating  `InitContainerStatus`.

euank commented:
Review status: 0 of 6 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 530 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSLPVZOlpG1huhvy-bF-r1-530:-KSi-9HxM5T4Hv36oLlN:bf0b4t) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_container.go#L530)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Given that we print `count` in the log message below, let's just keep it as it is. @euank are you ok with that? 
</details>
I did see that we printed it, but I don't think it adds much value in being printed. I'm fine with leaving it as is (just a minor readability nit)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363)*
<!-- Sent from Reviewable.io -->

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -560,6 +557,84 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *api.Pod, r
+// findActiveInitContainer returns the status of the last failed container, the next init container to
+// start, or done if there are no further init containers. Status is only returned if an init container
+// failed, in which case next will point to the current container.
+func findActiveInitContainer(pod *api.Pod, podStatus *kubecontainer.PodStatus) (next *api.Container, status *kubecontainer.ContainerStatus, done bool) {
```
The function name is confusing. It's returning the next-to-be-run init container, which can be a failed one, so why it's called active init containers?

feiskyer replied:
Good point, renamed it to `findNextInitContainer`.

yifan-gu commented:
TBH, I found this one hard to review quickly. Still haven't finished reviewing the `computePodContainerChanges()`

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -560,6 +557,84 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *api.Pod, r
+	}
+}
+
+// findActiveInitContainer returns the status of the last failed container, the next init container to
```
ack

feiskyer commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -380,14 +387,42 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *api.Pod, pod
+	// always reset the init containers if the sandbox is changed.
+	if !sandboxChanged {
+	Containers:
+		// keep all successfully completed containers up to and including the first failing container
```
ack

feiskyer commented:
@yujuhong @yifan-gu Renamed `findActiveInitContainer` to  `findNextInitContainer` and made a little refactor on it. PTAL.

euank commented:
Reviewed 3 of 5 files at r1, 1 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/kuberuntime/helpers.go, line 127 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSlr9hUFDPfbUSuFAU4:-KSlr9hUFDPfbUSuFAU5:b-bslh1i) ([raw file](https://github.com/kubernetes/kubernetes/blob/c52a714d27e69510b0e8209a4a942c6a5a1ac94b/pkg/kubelet/kuberuntime/helpers.go#L127)):*
> ```Go
> 
> func getContainerSpec(pod *api.Pod, containerName string) *api.Container {
> 	var containerSpec *api.Container
> ```

readability nit, you can nix this variable and just have returns where you currently have breaks. That would also let you get rid of the containerSpec == nil bit.

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 584 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSLPVZOlpG1huhvy-bF-r1-584:-KSmOpGMNujsLIl-QXUy:b-rhpbyw) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_container.go#L584)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> @euank @smarterclayton Made a little refactor, PTAL.
</details>
Looks clearer to me.

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 615 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSlsAnC5MjdVppcpHrQ:-KSlsAnC5MjdVppcpHrR:b-op6v1a) ([raw file](https://github.com/kubernetes/kubernetes/blob/c52a714d27e69510b0e8209a4a942c6a5a1ac94b/pkg/kubelet/kuberuntime/kuberuntime_container.go#L615)):*
> ```Go
> 		container := &pod.Spec.InitContainers[i]
> 		status := podStatus.FindContainerStatusByName(container.Name)
> 		if status != nil && status.State == kubecontainer.ContainerStateExited && status.ExitCode != 0 {
> ```

non-blocking nit, maybe we should have a helper function for this check, as suggested in this TODO: https://github.com/kubernetes/kubernetes/blob/c0826a2e7e9532da91c468dceb83790491f9f8e0/pkg/kubelet/dockertools/docker_manager.go#L1927

This check exists in 2 or 3 places at least I think.

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 414 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSLPVZOlpG1huhvy-bH-r1-414:-KSlu2Fp3Q-HJcArArZp:b-fsbbju) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L414)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> well, container should be kept event failed, so the logic will be same of changing it from `switch` to `if..else`
</details>
Other than not needing a break label for if, you're right. IMO it's worth changing for that reason alone, but I won't block if other people are fans of break-labels

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363)*
<!-- Sent from Reviewable.io -->

agmsb commented:
Please remove info@crowdfunder.com

On Fri, Sep 23, 2016 at 2:39 AM, Pengfei Ni <notifications@github.com>
wrote:

> This PR adds init containers support in CRI.
>
> CC @yujuhong <https://github.com/yujuhong> @Random-Liu
> <https://github.com/Random-Liu> @yifan-gu <https://github.com/yifan-gu>
>
> Also CC @kubernetes/sig-node
> <https://github.com/orgs/kubernetes/teams/sig-node>
> @kubernetes/sig-rktnetes
> <https://github.com/orgs/kubernetes/teams/sig-rktnetes>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/kubernetes/kubernetes/pull/33363
> Commit Summary
>
>    - CRI: Add init containers
>
> File Changes
>
>    - *M* pkg/kubelet/api/testing/fake_runtime_service.go
>    <https://github.com/kubernetes/kubernetes/pull/33363/files#diff-0>
>    (12)
>    - *M* pkg/kubelet/kuberuntime/kuberuntime_container.go
>    <https://github.com/kubernetes/kubernetes/pull/33363/files#diff-1>
>    (86)
>    - *M* pkg/kubelet/kuberuntime/kuberuntime_gc_test.go
>    <https://github.com/kubernetes/kubernetes/pull/33363/files#diff-2> (2)
>    - *M* pkg/kubelet/kuberuntime/kuberuntime_manager.go
>    <https://github.com/kubernetes/kubernetes/pull/33363/files#diff-3>
>    (135)
>    - *M* pkg/kubelet/kuberuntime/kuberuntime_manager_test.go
>    <https://github.com/kubernetes/kubernetes/pull/33363/files#diff-4>
>    (150)
>
> Patch Links:
>
>    - https://github.com/kubernetes/kubernetes/pull/33363.patch
>    - https://github.com/kubernetes/kubernetes/pull/33363.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33363>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFIM7o786jtkwypFNm5hT7WZtsjGFCDlks5qs55rgaJpZM4KEylX>
> .
>



-- 
Crowdfunder Team

Support small businesses. Create jobs. Help rebuild the American economy:
crowdfunder.com <http://www.crowdfunder.com/>

yujuhong commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 530 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSLPVZOlpG1huhvy-bF-r1-530:-KSn6lhhB1hkN2go4puI:b-c2v7z0) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_container.go#L530)):*
<details><summary><i>Previously, euank (Euan Kemp) wrote…</i></summary>
> I did see that we printed it, but I don't think it adds much value in being printed. I'm fine with leaving it as is (just a minor readability nit)
</details>
Given the init container code is relatively new, let's keep this for (potential) debuggability.

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 602 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSn7ThE9aMCsYr8tZ8k:-KSn7ThE9aMCsYr8tZ8l:b-ljr7cs) ([raw file](https://github.com/kubernetes/kubernetes/blob/53b8f07bab87864cc55dcb2c959b7adeb8d775e5/pkg/kubelet/kuberuntime/kuberuntime_container.go#L602)):*
> ```Go
> 				m.containerRefManager.ClearRef(status.ID)
> 			} else {
> 				glog.Warningf("No ref for container '%q'", status.ID)
> ```

nit: don't think we need single quotes for %q

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 573 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSLPVZOlpG1huhvy-bF-r3-573:-KSn9gp_2NEt12vf283B:bnovgms) ([raw file](https://github.com/kubernetes/kubernetes/blob/47c578b084734b32535ff7df74a2a38b8bcec63e/pkg/kubelet/kuberuntime/kuberuntime_container.go#L573)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> Not sure whether that could make things easier, same logic is required while populating  `InitContainerStatus`.
</details>
Yep. Probably not gonna help much, and it requires more information about the pod spec to populate `kubecontainer.Status`

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 606 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSLPVZOlpG1huhvy-bF-r3-606:-KSn9tG6gRHSDf_WFX7S:b-pbdakg) ([raw file](https://github.com/kubernetes/kubernetes/blob/47c578b084734b32535ff7df74a2a38b8bcec63e/pkg/kubelet/kuberuntime/kuberuntime_container.go#L606)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> Good point, renamed it to `findNextInitContainer`.
</details>
nit: `findNextInitContainerToRun` to be even more clear?

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 615 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSlsAnC5MjdVppcpHrQ:-KSn9Wvfiaelpczolsoz:b-p5j7jm) ([raw file](https://github.com/kubernetes/kubernetes/blob/c52a714d27e69510b0e8209a4a942c6a5a1ac94b/pkg/kubelet/kuberuntime/kuberuntime_container.go#L615)):*
<details><summary><i>Previously, euank (Euan Kemp) wrote…</i></summary>
> non-blocking nit, maybe we should have a helper function for this check, as suggested in this TODO: https://github.com/kubernetes/kubernetes/blob/c0826a2e7e9532da91c468dceb83790491f9f8e0/pkg/kubelet/dockertools/docker_manager.go#L1927
> 
> This check exists in 2 or 3 places at least I think.
</details>
+1 for adding a helper.

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 414 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSLPVZOlpG1huhvy-bH-r1-414:-KSn96BzE5QpDGwFgISe:bxu45xy) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L414)):*
<details><summary><i>Previously, euank (Euan Kemp) wrote…</i></summary>
> Other than not needing a break label for if, you're right. IMO it's worth changing for that reason alone, but I won't block if other people are fans of break-labels
</details>
not a big fan of break lables, unless it dramatically simplifies the code. If the alternative is straightforward, I'd go for it. This is non-blocking to me. 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 589 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSnDfoqcX0cJdXmZfyG:-KSnDfoqcX0cJdXmZfyH:b64adck) ([raw file](https://github.com/kubernetes/kubernetes/blob/c52a714d27e69510b0e8209a4a942c6a5a1ac94b/pkg/kubelet/kuberuntime/kuberuntime_container.go#L589)):*
> ```Go
> 			glog.V(4).Infof("Removing init container %q instance %q %d", status.Name, status.ID.ID, count)
> 			if err := m.runtimeService.RemoveContainer(status.ID.ID); err != nil {
> 				utilruntime.HandleError(fmt.Errorf("failed to remove pod init container %q: %v; Skipping pod %q", status.Name, err, format.Pod(pod)))
> ```

Why we don't need the `count--` logic here now?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 450 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KScIWbFLT6dtYtsrUDl:-KSngioDFePLdYfhYL4e:b-lrc4eo) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L450)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> I know you copied this from `docker_manager.go`, and there was probably a good reason for having this logic, I just couldn't remember why there would be containers running if initialization failed. 
</details>
One possibility I can imagine is that:
1. The pod was restarted
2. Some containers were not properly stopped, which could happen
3. New initialization run and failed
4. We got here...

I'm not sure whether this is the case we are handling here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 478 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KScFN4JvRvmDum_VatR:-KSnioXLAgeo16kDGWrH:boxa1oc) ([raw file](https://github.com/kubernetes/kubernetes/blob/30d586de5ccb6b27523a7dcc81ca0b329965c7f2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L478)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Why would this happen?
</details>
Same with last comment, when the pod is restarted but all containers are not needed to restart.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363)*
<!-- Sent from Reviewable.io -->

feiskyer commented:
> Please remove info@crowdfunder.com

@agmsb Not sure what's the problem. I didn't refer to this email anywhere. Could you clarify it clear?

> Why we don't need the count-- logic here now?

@Random-Liu  CRI requires `RemoveContainer` returns success if the container is not found.

@yujuhong @euank Addressed comments, and removed break labels by moving the logic to `checkAndKeepInitContainers`. PTAL.

Random-Liu commented:
> @Random-Liu CRI requires RemoveContainer returns success if the container is not found.

Ok. In fact that `count` is only used for log, so it should be fine. :)

euank commented:
I left a small nit, but LGTM either way.

---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 402 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-KSrTx2s4rYVxyo2w4q2:-KSrTx2s4rYVxyo2w4q3:b-rhklsb) ([raw file](https://github.com/kubernetes/kubernetes/blob/1e998ee6ba28b0721ca6782d76ed53d27061c5b9/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L402)):*
> ```Go
> 		if containerStatus.State == kubecontainer.ContainerStateExited {
> 			initContainersToKeep[containerStatus.ID] = i
> 			if containerStatus.ExitCode != 0 {
> ```

nit, we have a helper we could use here

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33363#-:-KSrUAemVylybF9knZx0:bggcarp)*
<!-- Sent from Reviewable.io -->

yifan-gu commented:
Taking another review.

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -374,20 +381,60 @@ func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *api.Pod, podStatus *k
+	}
+	changes.InitFailed = initFailed
+
+	// check the status of containers.
```
If init containers failed, why should we even bother check the Spec.Containers???

feiskyer replied:
There could. Refer https://github.com/kubernetes/kubernetes/pull/33363#issuecomment-250353318.

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -428,10 +488,17 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *api.Pod, pod
-	runningContainerStatues := podStatus.GetRunningContainerStatuses()
-	for _, containerStatus := range runningContainerStatues {
-		if _, keep := changes.ContainersToKeep[containerStatus.ID]; !keep {
+	// Don't keep init containers if they are the only containers to keep.
```
Under which scenario will this happen? Any examples?

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -406,6 +453,19 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *api.Pod, pod
+		if initFailed {
+			// Initialization failed and Container exists.
+			// If we have an initialization failure everything will be killed anyway.
+			// If RestartPolicy is Always or OnFailure we restart containers that were running before we
```
@feiskyer I can't parse this comment, could you please make it clearer?

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -585,7 +654,59 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *api.Pod, _ api.PodStatus, podSt
 
-	// Step 5: start containers in podContainerChanges.ContainersToStart.
+	// Step 5: start init containers.
+	status, next, done := findNextInitContainerToRun(pod, podStatus)
```
New function name looks fine, but I wonder why we need to do pod status again to know the next init container to start?
Shouldn't computeChanges() do the job?

feiskyer replied:
Because of `m.pruneInitContainersBeforeStart `.

yifan-gu commented:
I know it's copied from the old code, but I still feel it's making kubelet move towards *unmaintainable*. It's hard for the maintainers here (me at least) to confidently reason about the correctness of the logic. So we largely depend on the init e2e tests. I am concerned that if there are any bugs or races later, it will give us a hard time to fix.

Not blocking this so we can move on, but feel strongly we need to do a clean up after this, And I am happy to do that.

feiskyer commented:
@yifan-gu Agreed. But it is difficult to refactor only init containers logic without refactor the whole syncpod.  I think we can have this in first to have a logic-complete syncpod and re-consider refactoring later.

yifan-gu commented:
@feiskyer I am ok with the plan.

yujuhong commented:
@yifan-gu I totally agree that more cleanup would go a long way for maintenance in the future (and thanks for offering to do that!).
I am ok with this PR going it as it is, and refactoring the code later.

yujuhong commented:
PR LGTM, with refactoring in the followup PRs. 
Anyone still has comment for this PR?

yifan-gu commented:
@k8s-bot gke e2e test this #33617
